### PR TITLE
Fix dispatch of SET TIME ZONE INTERVAL statement.

### PR DESCRIPTION
**Pull request by soumyadeep2007**
This fixes Issues: #9055 and incidentally fixes related issue #9044.

Tl;dr: Instead of dispatching `SET TIME ZONE INTERVAL <args>` when this
flavor was used, we extracted the arguments following `INTERVAL` and
dispatched `SET timezone TO <args>`.

Please refer to issue #9055 for the repro, complete RCA and fix details.

Additional Notes:

1. While working on these two issues, I saw that `DispatchSetPGVariable()`
became out of date with respect to upstream code. This is because it has
a foreach loop to add the argument list to the dispatch string that is
borrowed heavily from `flatten_set_variable_args()`. In fact that is the
precise reason for related issue: #9044. We did not port the changes
brought by upstream commit: 6fff5c3. So I ported those changes over.

2. I added a test for this scenario and I refactored some existing tests
in this area for them to be more explicit in their intent.

### CODE REVIEW

File: src/backend/utils/misc/guc.c [hunk 1]
```diff
@@ -7723,58 +7725,22 @@ DispatchSetPGVariable(const char *name, List *args, bool is_local)
...
-
-
-					break;
-				default:
-					elog(ERROR, "unrecognized node type: %d",
-						 (int) nodeTag(&arg->val));
-					break;
-			}
-		}
+		if (strcmp(name, "timezone") == 0)
```
Comment (baobao0206) [inline src/backend/utils/misc/guc.c]: Why do we need to handle timezone separately?

Comment (soumyadeep2007) [reply src/backend/utils/misc/guc.c]: I will add a clarifying comment. The issue is that if we ever use `SET TIME ZONE INTERVAL <args>` syntax which is an alias for `SET timezone...` we should dispatch `SET TIME ZONE INTERVAL<args>` as opposed to the current behavior (`SET timezone TO <args>`).
So this is one way to fix the issue. Please refer: #9055 for more details.

File: src/backend/utils/misc/guc.c [hunk 2]
```diff
@@ -7723,58 +7725,22 @@ DispatchSetPGVariable(const char *name, List *args, bool is_local)
...
+			appendStringInfo(&buffer, "%s",
+							 quote_literal_cstr(GetConfigOptionByName("timezone",
+																	  NULL,
+																	  false)));
+		else
+			appendStringInfo(&buffer, "%s", quote_literal_cstr(arg_string));
```
Comment (soumyadeep2007) [inline src/backend/utils/misc/guc.c]: This line is causing problems. We should not be quoting everything.

File: src/backend/utils/misc/guc.c (hunk 1 above)
Comment (soumyadeep2007) [reply src/backend/utils/misc/guc.c]: Added comment.

**Comment by dgkimura** [comment]
Your thorough analysis helped make this much easier to follow.

Naive and probably unrealistic question, what if the QD and QE were in different timezones and user specified `SET TIME ZONE DEFAULT`?  If I followed correctly, your patch would cause behavior change in this scenario?

**Comment by soumyadeep2007** [comment]
> Naive and probably unrealistic question, what if the QD and QE were in different timezones and user specified `SET TIME ZONE DEFAULT`? If I followed correctly, your patch would cause behavior change in this scenario?

Ah great catch, indeed, you are spot on. Though I doubt that any user would have QEs in different time zones but this is a case worth accounting for. I would add some special case handling for `default`.

**Comment by soumyadeep2007** [comment]
Hmm special case handling will make the code unnecessarily complicated. 

Also, there is not much value in allowing the user to set the time zone to different default/local values for different QD/QEs anyway. One may argue that the time zone that is local/default for the QD should be propagated to the QEs. We should probably never have allowed this in the first place.

**Comment by dgkimura** [comment]
> Hmm special case handling will make the code unnecessarily complicated.
> 
> Also, there is not much value in allowing the user to set the time zone to different default/local values for different QD/QEs anyway. One may argue that the time zone that is local/default for the QD should be propagated to the QEs. We should probably never have allowed this in the first place.

Given those arguments, it seems reasonable to me to ignore.

File: src/backend/utils/misc/guc.c [hunk 3]
```diff
@@ -7735,44 +7735,69 @@ DispatchSetPGVariable(const char *name, List *args, bool is_local)
...
 		appendStringInfo(&buffer, "%s TO ", quote_identifier(name));
 
-		foreach(l, args)
+		/*
+		 * GPDB: We handle the timezone GUC specially. This is because the
+		 * timezone GUC can be set with the SET TIME ZONE .. syntax which is an
+		 * alias for SET timezone. Instead of dispatching the SET TIME ZONE ..
+		 * as a special case, we dispatch the already set time zone from the QD
+		 * with the usual SET syntax flavor (SET timezone TO <>).
+		 * Please refer to Issue: #9055 for additional detail.
```
Comment (dgkimura) [inline src/backend/utils/misc/guc.c]: Nit: I haven't seem many issue # in the code.  And after a quick grep, those that are there don't seem to be github issues.  I imagine this might be confusing for a future archeologist.

Comment (soumyadeep2007) [reply src/backend/utils/misc/guc.c]: I agree, I included the URL for better clarity. I think we should encourage issue numbers more, specially if they contain extra detail that would otherwise bloat the commit message or comment(s) referencing it.

**Comment by soumyadeep2007** [comment]
Thanks for the review!
